### ISSUE TITLE
ag slower than find, xargs, grep with high CPU count

### ISSUE THREAD

**Issue by zachriggle** [ggreer/the_silver_searcher#608]
## Test System

```shell
$ cat /proc/cpuinfo | grep Xeon | uniq -c
     40 model name      : Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz
$ free -h
             total       used       free     shared    buffers     cached
Mem:           62G        52G        10G       283M       3.3G        37G
-/+ buffers/cache:        11G        51G
Swap:          63G        24M        63G
$ sudo hdparm -I /dev/sda | grep Model
        Model Number:       Samsung SSD 850 PRO 1TB 
```

## Test Corpus

This is a checkout of the Android source tree.

```shell
$ find . -name .repo -prune -o -name .git -prune -o -name out -prune -o -type f | wc -l
545561
```

## Competing Command

The `jgrep` command comes from Android's [envsetup.sh](https://github.com/android/platform_build/blob/master/envsetup.sh) script.  It's just `find`, `xargs`, and `grep`.

```shell
#!/bin/bash
jgrep ()
{
        find . -name .repo -prune -o -name .git -prune -o -name out -prune -o -type f -name "*\.java" -print0 | xargs -0 grep --color -n "$@"
}

jgrep Android | wc -l
```

```shell
$ time bash jgrep-example.sh
39542
bash jgrep-example.sh  2.21s user 1.31s system 143% cpu 2.455 total
```

## `ag` 0.29.1 performance

With `~/.agignore`:

```
.repo
.git
out
```

Ag 0.29.1 takes about 3x as long to perform the search.

```shell
$ ag -V
ag version 0.29.1
$ time ag --java Android | wc -l
38743
ag --java Android  5.45s user 7.07s system 192% cpu 6.507 total
wc -l  0.00s user 0.03s system 0% cpu 6.507 total
```

It's roughly the same with no `.agignore`:

```shell
$ rm -f ~/.agignore
$ time ag --java Android | wc -l
38743
ag --java Android  5.28s user 7.24s system 196% cpu 6.379 total
wc -l  0.01s user 0.03s system 0% cpu 6.378 total
```

And no improvement for string literals.

```shell
$ time ag -Q --java Android | wc -l
38743
ag -Q --java Android  5.39s user 7.48s system 193% cpu 6.659 total
wc -l  0.01s user 0.03s system 0% cpu 6.659 total
```

## `ag` 0.19.2 performance

Ag 0.19.2 (which comes with Ubuntu 14.04.1) isn't as bad on performance.  However, it does miss about 25% of the lines which it should find.

```shell
$ ag.old -V
ag version 0.19.2
$ time ag.old --java Android | wc -l
30785
ag.old --java Android  3.27s user 3.52s system 173% cpu 3.914 total
wc -l  0.00s user 0.01s system 0% cpu 3.913 total
```

**Comment by ggreer** [comment]
Did you actually build the Android source or is this a clean checkout?

**Comment by zachriggle** [comment]
- Clean source tree
- Results were comparable when searching for non-existent strings (I_DONT_EXIST vs Android)
- Valid point for the comparison I posted, but not relevant if no lines are printed
- Also a valid point. Is there a way to reconcile the behavior for comparison?
- I agree that by default ag is fast and easy, but it's selling point is that it's fast*er*.  This doesn't seem to hold true against older versions of itself or grep.

**Comment by zachriggle** [comment]
Here's something with more standard hardware -- the most recent 4-core/8-thread i7 Retina Macbook Pro.  After prepping the cache by running a few times until I got consistent results, this is what the numbers look like.

Of course this is now on a different OS (Ubuntu 14.04 above vs Mac OS X Yosemite here), but the thread count has been cut from 40 to 8, and memory reduced from 64GB to 16GB.

```
$ fdesetup status
FileVault is On.
$ system_profiler SPHardwareDataType
Hardware:

    Hardware Overview:

      Model Name: MacBook Pro
      Model Identifier: MacBookPro11,2
      Processor Name: Intel Core i7
      Processor Speed: 2.2 GHz
      Number of Processors: 1
      Total Number of Cores: 4
      L2 Cache (per Core): 256 KB
      L3 Cache: 6 MB
      Memory: 16 GB
...
$ system_profiler SPSerialATADataType
SATA/SATA Express:

    Apple SSD Controller:

      Vendor: Apple
      Product: SSD Controller
      Physical Interconnect: PCI
      Link Width: x2
      Link Speed: 5.0 GT/s
      Description: AHCI Version 1.30 Supported

        APPLE SSD SM0256F:

          Capacity: 251 GB (251,000,193,024 bytes)
          Model: APPLE SSD SM0256F
```

### jgrep

Exact same script as before.

```
$ time bash jgrep-example.sh
   32497
bash jgrep-example.sh  5.52s user 19.27s system 82% cpu 30.194 total
```

### ag 0.29.1

Much faster than jgrep here.

```
$ ~/the_silver_searcher/ag --version
ag version 0.29.1
$ time ~/the_silver_searcher/ag --java Android | wc -l
   32692
~/the_silver_searcher/ag --java Android  2.71s user 4.19s system 93% cpu 7.346 total
wc -l  0.01s user 0.00s system 0% cpu 7.345 total
```

### ag 0.19.2

I also built this version for testing against the results above.  Looks like the most recent release is indeed faster.

```
$ ~/the_silver_searcher/ag --version
ag version 0.19.2
$ time ~/the_silver_searcher/ag --java Android | wc -l
   32696
~/the_silver_searcher/ag --java Android  3.49s user 4.58s system 92% cpu 8.743 total
wc -l  0.01s user 0.00s system 0% cpu 8.742 total
```

**Comment by zachriggle** [comment]
On further consideration to the points you raised;

1. The clean- or not-clean working directory doesn't have a significant impact.
2. Printing is not the bottleneck as previously stated, for example if we search for a non-existent phrase, DOES_NOT_EXIST, versus a common one, Android.
3. In my benchmark example, grep *is* printing out line numbers (`-n`).
4. All of the `grep` and `jgrep` output is correctly serialized.  `xargs` does not actually ever execute multiple instances of grep in parallel.  It *would* do this if `-P` is specified, but it does not.
5. I still don't disagree, but clearly this is a feature that `ag` is supposed to handle gracefully.  The existence of a `--java` flag and appreciation for `.agignore` files is evidence of this.

### Rectified Linux Tests

Also, here's a rectified benchmark to address the concerns about the `print` code being slower, or about the heading groupings being an issue.  This is back on the 40-core/64GB Ubuntu machine.  I also threw one in to manually set the worker count.

```
$ ag --version
ag version 0.29.0
$ time ag --nogroup --java I_DONT_EXIST --stats
0 matches
49695 files searched
402353855 bytes searched
4.738296 seconds
ag --nogroup --java I_DONT_EXIST --stats  3.17s user 4.62s system 164% cpu 4.740 total
```

```
$ time ag.old --nogroup --java I_DONT_EXIST --stats
0 matches
42259 files searched
336320422 bytes searched
3.970165 seconds
ag.old --nogroup --java I_DONT_EXIST --stats  2.62s user 3.58s system 156% cpu 3.972 total
$ ag.old --version
ag version 0.19.2
```

### Restricted Worker Counts

The performance is about 12% faster when only using 1 CPU core.

```
$ time ag --workers 1 --java STRING_DOESNT_EXIST
ag --workers 1 --java STRING_DOESNT_EXIST  2.59s user 2.01s system 119% cpu 3.870 total
$ time ag.old --workers 1 --java STRING_DOESNT_EXIST
ag.old --workers 1 --java STRING_DOESNT_EXIST  2.13s user 1.89s system 117% cpu 3.421 total
```

The performance only goes lower the more workers are introduced.

```
$ time ag --workers 2 --java STRING_DOESNT_EXIST
ag --workers 2 --java STRING_DOESNT_EXIST  2.70s user 2.45s system 132% cpu 3.902 total
$ time ag --workers 8 --java STRING_DOESNT_EXIST
ag --workers 8 --java STRING_DOESNT_EXIST  2.97s user 3.14s system 151% cpu 4.042 total
$ time ag --workers 16 --java STRING_DOESNT_EXIST
ag --workers 16 --java STRING_DOESNT_EXIST  2.82s user 3.59s system 152% cpu 4.191 total
$ time ag --workers 40 --java STRING_DOESNT_EXIST
ag --workers 40 --java STRING_DOESNT_EXIST  3.17s user 4.53s system 155% cpu 4.950 total
```
